### ISSUE TITLE
AnimatedBuilder should be renamed 

### ISSUE THREAD

**Issue by rrousselGit** [flutter/flutter#19255]
The class [`AnimatedBuilder`](https://docs.flutter.io/flutter/widgets/AnimatedBuilder-class.html) should have its name replaced. As it is not related to animation in reality.

For example, the following is possible :

```dart
final valueNotifier = ValueNotifier(0);

AnimatedBuilder(
  animation: valueNotifier,
  builder: (context, _) {
    return Text(valueNotifier.value.toString());
  },
)
```

Same goes for `ScrollController`, ...

I suggest to rename it to `ListenableBuilder` as it makes more sense.

**Comment by zoechi** [comment]
@Hixie

**Comment by rrousselGit** [comment]
We can raise the same question about `AnimatedWidget`. Ultimately it faces the same problem. 
Although `AnimatedWidget` is more honest about its behavior. As `AnimatedWidget` has its parameter named `listenable`. While `AnimatedBuilder` named it `animation` instead.


____

These names would be fine if the parameter had for type `Animation<T>` instead of `Listenable`. 

Maybe an alternative is to keep `AnimatedBuilder/Widget` but with the type `Animation<T>`. And change the builder to 
```
Widget animationBuilder(BuildContext context, Widget chid, T value);
```

And then offer a `ListenableBuilder/Widget` as a replacement to the current API.

**Comment by Hixie** [comment]
AnimatedWidget used to take an Animation, we changed it to a Listenable much later in its life (when it was first invented, Listenable didn't exist).

Changing this would be a very breaking change, it's probably too late to really make changes here. But it's something we can study when we look at the animation API updates we've discussed.

**Comment by rrousselGit** [comment]
A bit off topic but where can I find that "animation API update discussion"?

**Comment by xster** [comment]
I'm assuming you linked the wrong PR and are probably referring to #19729. I originally wanted to extract some base class out of the AnimatedBuilder but I think there's something inconvenient in the existing API (where the listable wasn't passed back into the builder closure signature). Since a ValueListenableBuilder is as much an encapsulation mechanism as a triggering mechanism, it would be inappropriate for its owner to pass the listenable back into the builder function instead of being self sufficient. 

TL;DR, we're unlikely to rename the AnimatedBuilder and you can consider using the ValueListenableBuilder once the PR is merged.

**Comment by rrousselGit** [comment]
That's already a pretty good start, nice !

In the continuation of this PR, would it be valid if I contributed with an `ListenableBuilder` and `ScrollControllerBuilder` ?

_____

Back on the topic, can't we create an Animat**ion**Builder and mark Animat**ed**Builder as `@deprecated` with other the new builders listed as alternatives ?

Animat*ionBuilder would be fine as name considering it would listen `Animation`. Afterall we don't have a `StreamedBuilder` but `SteamBuilder`.

**Comment by xster** [comment]
Ha, we actually made one too 
https://github.com/flutter/flutter/pull/18136 (for scrolling), though we didn't check it in because we weren't sure if people would use it. 

Regarding the existing API, like Hixie said, we're unfortunately unlikely to change it.

**Comment by rrousselGit** [comment]
Regarding the scroll builder, I was thinking of listening `ScrollController` instead of notifications.

I used this pattern a lot for things like parallax or when I need to align an overlay next to a scroll item. Which wouldn't be possible with notifications.

**Comment by escamoteur** [comment]
@Hixie indeed it's a problem that people don't realise that AnimatedBuilder can just take any ListenableBuilder. They try to use ValueListenableBuilder and fail because it doesn't accept any Listenable but only Value Changenotifier.
Couldn't we just add a ListenableBuilder as alternative to AnimatedBuilder. I know it's stupid to have the same functionality twice but users would know which one to use with Listenables.

**Comment by escamoteur** [comment]
IMHO we should discuss what possibilities we have here. It's really difficult for beginners to understand that an animation just triggers Listenables and therefore the AnimationBuilder can be used with any Listenable

**Comment by dungnv2602** [comment]
Please Flutter team, by any means I can't swallow this, it keeps annoying me all the time cause it sounds horribly wrong.
Stream => StreamBuilder
ValueListenable => ValueListenableBuilder
Listenable => AnimatedBuilder

**Comment by ozburo** [comment]
Well, whether or not you guys decide to rename this Class or not, I learned something today 👍 

Now I don't need to populate my initState and setState with tons of ScrollController offsets/etc and simply delegate that to AnimatedBuilder for each of my Widgets that need them.

IMO the really confusion is around the member name `animation` which implies you need an instance of AnimationController and no *any* controller -- perhaps this is what needs to be changed or made more "generic".

In any case thanks for teaching me something, I feel this a level up in terms of mastering Flutter and should be required reading 💯

**Comment by icnahom** [comment]
> While at it, a "plural" version of this would be quite nice as well. Sometimes you have multiple listenables, changing any of them requires a rebuild, nice not to have to nest builders at the same pt in the tree.

Try passing them to Listenable.merge(), and use this listenable in an AnimatedBuilder.

And it would be great if the team accepts your suggestions about slowing moving to ListenableBuilder.The new skeleton app is using AnimatedBuilder and I don't think it's a good name.

**Comment by mateusfccp** [comment]
Yeah, I think this is totally doable now. We could simply introduce this:

```dart
@Deprecated('Use ListenableBuilder and bla bla bla ')
typedef AnimatedBuilder = ListenableBuilder;
```

And then we remove it completely in a later stable version.

I don't think it's much of a hassle...

**Comment by albertodev01** [comment]
There is a new PR (https://github.com/flutter/flutter/pull/97256) to try to solve this issue

**Comment by esDotDev** [comment]
Just to kick the tires on this a bit, https://github.com/flutter/flutter/issues/95967#issuecomment-1003238756

>> Actually, I'm a bit unsure if we want to create an alias for an existing API (AnimatedBuilder), just for the sake of it being more discoverable. This might confuse API users as well - why are there two classes doing the same thing?

It would not primarily be to make it more discoverable (though that is a side-benefit),  it would be so we can write proper semantic code that reads correctly and does not mis-lead the reader. 

The rename would not really be confusing if `AnimatedBuilder` is deprecated, and eventually removed after a few yrs.
